libcnb/src/layer/test.rs

```diff
@@ -0,0 +1,792 @@
+//! Tests for [`Layer`] and [`layer::handle_layer`].
+//!
+//! Even though they are implemented as unit tests, they are really integration tests since they
+//! need files on disk, an implementation of [`Layer`] and [`Buildpack`] and test logic across all
+//! of these parts. There is no other way of testing this if we want to keep the public API as-is.
+//! Please note that individual building blocks that are used in [`layer::handle_layer`] are unit
+//! tested separately.
+//!
+//! All tests in this module assume a specific [`Layer`] implementation that is also in this file.
+//! Look for the `TestLayer` type and it's [`Layer`] implementation.
+
+use crate::build::{BuildContext, BuildResult, BuildResultBuilder};
+use crate::data::buildpack_id;
+use crate::data::layer_content_metadata::LayerTypes;
+use crate::data::stack_id;
+use crate::detect::{DetectContext, DetectResult, DetectResultBuilder};
+use crate::generic::{GenericMetadata, GenericPlatform};
+use crate::layer::{
+    handle_layer, ExistingLayerStrategy, Layer, LayerData, LayerResult, LayerResultBuilder,
+    MetadataMigration,
+};
+use crate::layer_env::{LayerEnv, ModificationBehavior, TargetLifecycle};
+use crate::{read_toml_file, Buildpack, Env, LIBCNB_SUPPORTED_BUILDPACK_API};
+use libcnb_data::buildpack::{BuildpackToml, Stack};
+use libcnb_data::buildpack_plan::BuildpackPlan;
+use libcnb_data::layer::LayerName;
+use libcnb_data::layer_content_metadata::LayerContentMetadata;
+use rand::Rng;
+use serde::Deserialize;
+use serde::Serialize;
+use std::fs;
+use std::path::Path;
+use tempfile::{tempdir, TempDir};
+
+const TEST_LAYER_LAUNCH: bool = true;
+const TEST_LAYER_BUILD: bool = true;
+const TEST_LAYER_CACHE: bool = true;
+const TEST_LAYER_CREATE_FILE_CONTENTS: &str = "ran";
+const TEST_LAYER_UPDATE_FILE_CONTENTS: &str = "ran";
+const TEST_LAYER_CREATE_FILE_NAME: &str = "create";
+const TEST_LAYER_UPDATE_FILE_NAME: &str = "update";
+
+#[derive(Clone)]
+struct TestLayer {
+    existing_layer_strategy: ExistingLayerStrategy,
+    write_layer_env: Option<LayerEnv>,
+    write_version: String,
+}
+
+impl Layer for TestLayer {
```

Comment:
Do we need a few more tests that also test a `Layer` that only has implementations for the mandatory methods, so we can test the default implementations for these?
- `existing_layer_strategy()`
- `update()`
- `migrate_incompatible_metadata()`

Reply:
Hard to argue against it when we had an actual defect previously :) I'll see how these can be implement without too much boilerplate. Ideally, we just test the specific implementations and not end-to-end.

Reply:
Added in 5599990